### DISCUSSION TITLE
Support for multiple blank lines

### DISCUSSION THREAD

**Discussion by puopg** [remarkjs/remark#993] [Ideas] [answered]
This is more of a discussion around approaches toward handling multiple blank lines in sequence. I am definitely not knowledgeable on proper markdown syntax, and what may or may not be an ok thing to do, but I was hoping to get enlightened on an approach that seems to work toward adding support for multiple blank lines

The example use case would be akin to things like discord and slack, where one can insert multiple newlines into a chat message and they are honored in the rendered output.

A markdown string with multiple sequential newlines would be something like: 

`Hello World, this is line 1\n\n\n\nAnd this is line 6`

An approach I looked to using actually doesn't involve the MDAST, or even micromark extensions, but instead, ever so slightly modifying the input markdown string. Honestly, I've tried to understand why newlines get collapsed digging through the micromark code base, but Im afraid my understanding of what is going on there just isn't there yet.

Anyways, so the idea I had was to just replace line breaks (LF, CR, CRLF) with a newline + non breaking space character. It seems with this, in conjunction with `remark-breaks`, I can achieve the behavior that I wanted. As for why it happens, I want to say its because there is now content (the nbsp) before the newline, and remark-breaks handles that brilliantly. Not sure though 100%.

I would love some input into this approach, reasons why the approach may run into problems. And I would love to also get some feedback on if behavior like this would best be handled at the micromark level, MDAST level, or elsewhere.

Anyways, here is a small demo of this. 

#### Using

`react-markdown`: `8.0.3`
`remark-breaks`: `3.0.2`

## Demo of approach
https://codesandbox.io/s/create-react-app-forked-rfn632?file=/src/App.js

**Comment by wooorm** [comment, marked as answer]
Hey!

I believe this goes against the fundamental *nature* of markdown (and HTML), which marks up things to have meaning, and is not about WYSIWYG “drawing” of things.
Parsing markdown (and HTML), following standards, is also extremely involved: it’s super complex, there are all these edge cases that are handled by micromark/remark/etc.
If you just want to do something that discord or slack also do, and a bit different, then you’re not helped by markdown. You’re looking for something else. A couple of regexes. Something like https://github.com/developit/snarkdown maybe. There are probably more projects that slightly enhance comment boxes.
On the other hand, if you do want a standard that is supported in around the web, then do use markdown, without such every-line-ending-turns-into-a-visible-line-ending.

So yeah, you can regex find-and-replace line endings and add `&nbsp;`s there as a “patch”. But you probably end up spending more time fighting with markdown afterwards.

**Reply by puopg** [reply]
Gotcha, thanks for the insight! Yea I mean I totally get what you mean by just big edge cases haha. 

As far as use case goes, id love to know if there are similar to react markdown alternatives for parsing markdown and presenting the developer with a nice component interface to render with? I think that’s one of the main reasons I liked react markdown since it’s so easy to render stuff and configure to my needs. Plus the added benefit to have remark plugin support was pretty sweet.


On the topic of new lines, regarding the “fighting with markdown” comment, could you give an example of what type of battle you mean? Since I’m my head, I would still be giving valid markdown to the parser I think?

**Reply by wooorm** [reply]
I don’t know alternatives, no!

Some quick problems:

- Line endings can occur in non-“prose” places, e.g., also in (fenced) code, where character references don’t work
- Blank lines are sometimes needed in markdown: it changes how stuff is rendered

**Comment by puopg** [comment]
Another alternative approach here using micromark's `lineEndingBlank`. Any issues with this approach?

Totally get that the nodes rendered is no longer compliant with the commonmark spec, but if thats not a concern to peoples use case, is this ok to do? Or rather, is there an alarming issue that jumps out to you? 

Basically, the approach was to use an mdast extension to allow lineEndingBlanks to be represented in the final mdast. 

### Demo of alternative approach

https://codesandbox.io/s/create-react-app-forked-t6vzzj?file=/src/App.js

**Reply by wooorm** [reply]
I wholeheartedly hope that people do not throw specifications out of the window. As an example, companies like Microsoft/GitHub “doing their own thing” have made everyones lives, and significantly mine, much harder.

Yes, there might be issues with an approach of working with events/tokens. You are working on a complex, low, level, that you might not need to work on (combined with `react-markdown` which is super high-level?). There can be good reasons to work here, but there are many reasons not to work there.
There are also some issues with adding custom nodes into a tree. Though, there can be many very good reasons to do this.

ASTs have positional info. You might be able to use positional info to infer how many lines there were between things.
